server/api/models/User.go

```diff
@@ -0,0 +1,174 @@
+package models
+
+import (
+	"errors"
+	"html"
+	"log"
+	"strings"
+	"time"
+
+	"github.com/badoux/checkmail"
+	"github.com/jinzhu/gorm"
+	"golang.org/x/crypto/bcrypt"
+)
+
+type User struct {
+	ID uint32 `gorm:"primary_key;auto_increment" json:"id"`
+	Username string `gorm:"size:100;not null;unique" json:"username"`
+	Email string `gorm:"size:254;not null;unique" json:"email"`
+	Password string `gorm:"size:100;not null" json:"password"`
+	CreatedAt time.Time `gorm:"default:CURRENT_TIMESTAMP" json:"created_at"`
+	UpdatedAt time.Time `gorm:"default:CURRENT_TIMESTAMP" json:"updated_at"`
+}
+
+// Password handling
+func HashPassword(password string) (string, error) {
+	bytes, err := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)
+	return string(bytes), err
+}
+
+func VerifyPassword(password, hash string) error {
+	return bcrypt.CompareHashAndPassword([]byte(hash), []byte(password))
+}
+
+func (u *User) SaveHashedPassword() error {
+	hashedPassword, err := HashPassword(u.Password)
+	if err != nil {
+		return err
+	}
+	u.Password = string(hashedPassword)
+	return nil
+}
+
+func (u *User) SetUpUser() {
+	u.ID = 0
+	u.Username = html.EscapeString(strings.TrimSpace(u.Username))
+	u.Email = html.EscapeString(strings.TrimSpace(u.Email))
+	u.CreatedAt = time.Now()
+	u.UpdatedAt = time.Now()
+}
+
+func (u *User) Validate(action string) error {
```

Comment:
Would be better to split into two functions, I noticed the default and update have the same cases.